src/dhcp_proxy/dhcp_service.rs

```diff
@@ -149,50 +173,53 @@ impl From<DhcpServiceError> for Status {
-                        || old_lease.gateways != lease.gateways
-                    {
-                        // ips do not match, remove old ones and assign new ones.
-                        log::info!(
-                            "ip or gateway for mac {} changed, update address",
-                            &client.network_config.container_mac_addr
-                        );
-                        match update_lease_ip(
-                            &client.network_config.ns_path,
-                            &client.network_config.container_iface,
-                            old_lease,
-                            &lease,
-                        ) {
-                            Ok(_) => {}
-                            Err(err) => {
+pub async fn process_client_stream(service_arc: Arc<Mutex<DhcpV4Service>>) {
+    loop {
+        let lease_result = service_arc.lock().await.client.next().await;
+
+        if let Some(lease_result) = lease_result {
+            let mut client = service_arc.lock().await;
+            match lease_result {
+                Ok(lease) => {
+                    log::info!(
+                        "got new lease for mac {}: {:?}",
+                        &client.network_config.container_mac_addr,
+                        &lease
+                    );
+
+                    if let Some(old_lease) = &client.previous_lease {
+                        if old_lease.yiaddr != lease.yiaddr
+                            || old_lease.subnet_mask != lease.subnet_mask
+                            || old_lease.gateways != lease.gateways
+                        {
+                            log::info!(
+                                "ip or gateway for mac {} changed, update address",
+                                &client.network_config.container_mac_addr
+                            );
+                            if let Err(err) = update_lease_ip(
+                                &client.network_config.ns_path,
+                                &client.network_config.container_iface,
+                                old_lease,
+                                &lease,
+                            ) {
                                 log::error!("{err}");
-                                continue;
                             }
                         }
                     }
+                    client.previous_lease = Some(lease);
+                }
+                Err(err) => {
+                    log::error!(
+                        "Failed to renew lease for {}: {err}",
+                        &client.network_config.container_mac_addr
+                    );
                 }
-                client.previous_lease = Some(lease)
             }
-            Err(err) => log::error!(
-                "Failed to renew lease for {}: {err}",
-                &client.network_config.container_mac_addr
-            ),
+        } else {
+            break;
```

Comment:
this orphaned `break` always looks funny to me.  I'm not sure if you can envision a way to make this more readable.  nit only.